File: Python/shapes/rectangle.py
from shapes.shape import *


class RectangleXY(Shape):
    def __init__(self, x0, x1, y0, y1, z, material):
        self.x0 = x0
        self.x1 = x1
        self.y0 = y0
        self.y1 = y1
        self.z = z
        self.material = material

    def hit(self, r):
        t = (self.z - r.origin.z) / r.direction.z
        hit_pos = r.origin + (r.direction * t)
        if hit_pos.x < self.x1 and hit_pos.x > self.x0 and hit_pos.y > self.y0 and hit_pos.y < self.y1:
            return t
        return None

    def normal_at(self, pos):
        # return (pos - self.center).normalize()
        return Vector(z=1)


class RectangleXZ(Shape):
    def __init__(self, x0, x1, z0, z1, y, material):
        self.x0 = x0
        self.x1 = x1
        self.z0 = min(z0, z1)
        self.z1 = max(z0, z1)
        self.y = y
        self.material = material

    def hit(self, r):
        t = (self.y - r.origin.y) / r.direction.y
        hit_pos = r.origin + (r.direction * t)
        if hit_pos.x < self.x1 and hit_pos.x > self.x0 and hit_pos.z > self.z0 and hit_pos.z < self.z1:
            return t
        return None

    def normal_at(self, pos):
        # return (pos - self.center).normalize()
        return Vector(y=-1)


class RectangleYZ(Shape):
    def __init__(self, y0, y1, z0, z1, x, material):
        self.y0 = y0
        self.y1 = y1
        self.z0 = min(z0, z1)
        self.z1 = max(z0, z1)
        self.x = x
        self.material = material

    def hit(self, r):
        t = (self.x - r.origin.x) / r.direction.x
        hit_pos = r.origin + (r.direction * t)
        if hit_pos.z < self.z1 and hit_pos.z > self.z0 and hit_pos.y > self.y0 and hit_pos.y < self.y1:
            return t
        return None

    def normal_at(self, pos):
        # return (pos - self.center).normalize()
        return Vector(x=-1)

File: shapes/rectangle.cpp
#include "rectangle.h"

RectangleXY::RectangleXY(const double x0, const double x1, const double y0,
                         const double y1, const double z,
                         const std::shared_ptr<Material> &material, bool flipNormal)
    : Shape(material), x0(x0), x1(x1), y0(y0), y1(y1), z(z), flipNormal(flipNormal) {}

double RectangleXY::hit(const Ray &ray) const {
    double t = (z - ray.origin.z) / ray.direction.z;
    if (t < 1e-3) {
        return -1;
    }
    double x = ray.origin.x + t * ray.direction.x;
    double y = ray.origin.y + t * ray.direction.y;
    if (x < x0 || x > x1 || y < y0 || y > y1) {
        return -1;
    }

    // check if direction of ray and normal of rectangle are in the same direction
    // if (ray.direction.z * (flipNormal ? -1 : 1) < 0) {
    //     return -1;
    // }
    return t;
}

Vector RectangleXY::normalAt(const Vector &p) const { return Vector(0, 0, flipNormal ? -1 : 1); }

RectangleXZ::RectangleXZ(const double x0, const double x1, const double z0,
                         const double z1, const double y,
                         const std::shared_ptr<Material> &material, bool flipNormal)
    : Shape(material), x0(x0), x1(x1), z0(z0), z1(z1), y(y), flipNormal(flipNormal) {}

double RectangleXZ::hit(const Ray &ray) const {
    double t = (y - ray.origin.y) / ray.direction.y;
    if (t < 1e-3) {
        return -1;
    }
    double x = ray.origin.x + t * ray.direction.x;
    double z = ray.origin.z + t * ray.direction.z;
    if (x < x0 || x > x1 || z < z0 || z > z1) {
        return -1;
    }
    // check if direction of ray and normal of rectangle are in the same direction
    // if (ray.direction.y * (flipNormal ? -1 : 1) < 0) {
    //     return -1;
    // }

    return t;
}

Vector RectangleXZ::normalAt(const Vector &p) const { return Vector(0, flipNormal ? -1 : 1, 0); }

RectangleYZ::RectangleYZ(const double y0, const double y1, const double z0,
                         const double z1, const double x,
                         const std::shared_ptr<Material> &material, bool flipNormal)
    : Shape(material), y0(y0), y1(y1), z0(z0), z1(z1), x(x), flipNormal(flipNormal) {}

double RectangleYZ::hit(const Ray &ray) const {
    double t = (x - ray.origin.x) / ray.direction.x;
    if (t < 1e-3) {
        return -1;
    }
    double y = ray.origin.y + t * ray.direction.y;
    double z = ray.origin.z + t * ray.direction.z;
    if (y < y0 || y > y1 || z < z0 || z > z1) {
        return -1;
    }

    // check if direction of ray and normal of rectangle are in the same direction
    // if (ray.direction.x * (flipNormal ? -1 : 1) < 0) {
    //     return -1;
    // }
    return t;
}

Vector RectangleYZ::normalAt(const Vector &p) const { return Vector(flipNormal ? -1 : 1, 0, 0); }